Template Pipeline: Generate Temporary Variables

dylhunn opened the pull request:
The expression `a()?.b` should expand into `(tmp = a()) === null ? null : tmp.b`, in order to avoid calling the function `a()` twice.

This commit modifies the null-safe-expansion algorithm to emit temporary assignments, and provides the reification code to actually generate the declarations, assignments, and reads.

Note also that, with our bottom-up algorithm, there are some tricky cases when a function call exists inside an indexed access, such as `f1()?.[f2()?.a]?.b`. We add some special logic to avoid generating a double-assignment to the temporary storing the result of `f2()`.

Finally, there are opportunities to reuse the same temporary in expressions like `a?.[f()]?.[f()]`. We save this for the next commit.

eneajaho commented:
I have a question: 

What happens if the function has side effects? Or doesn't always return the same value. 

Nit: This is an absurd case, but it's a case 😅

dylhunn commented:
> I have a question:
> 
> What happens if the function has side effects? Or doesn't always return the same value.
> 
> Nit: This is an absurd case, but it's a case 😅

@eneajaho Due to this transformation, we call the function exactly once, which is what you might expect if you write `a()?.b`. So the behavior should not change. (Without this transformation, we'd cause the side effects to happen twice.)

dylhunn committed:
refactor(compiler): Generate temporary variable assignments when function calls appear in a safe-access expression.

The expression `a()?.b` should expand into `(tmp = a()) === null ? null : tmp.b`, in order to avoid calling the function `a()` twice.

This commit modifies the null-safe-expansion algorithm to emit temporary assignments, and provides the reification code to actually generate the declarations, assignments, and reads.

Note also that, with our bottom-up algorithm, there are some tricky cases when a function call exists inside an indexed access, such as `f1()?.[f2()?.a]?.b`. We add some special logic to avoid generating a double-assignment to the temporary storing the result of `f2()`.

Finally, there are opportunities to reuse the same temporary in expressions like `a?.[f()]?.[f()]`. We save this for the next commit.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
+    // TODO: Is this the right way to detect a pipe in the expression?
+    return true;
+  }
+  return false;
```
We should not have default behaviors for expression types - let's list everything out.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/reify.ts:
```diff
@@ -172,6 +172,16 @@ function reifyIrExpression(expr: o.Expression): o.Expression {
+      if (expr.name === null) {
+        throw new Error(`Assign of unnamed temporary ${expr.xref}`);
+      }
+      return new o.WriteVarExpr(expr.name, expr.expr);
```
```suggestion
      o.variable(expr.name).set(expr.expr);
```

alxhub commented on packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
 import * as ir from '../../ir';
 import {ComponentCompilation} from '../compilation';
 
+type SafeTransformContext = {
```
```suggestion
interface SafeTransformContext {
```

alxhub commented on packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
+  return e;
+}
+
+function maybeWithTemporary(e: o.Expression, ctx: SafeTransformContext) {
```
Declare return type

alxhub commented on packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
+    return [new ir.AssignTemporaryExpr(e, xref), new ir.ReadTemporaryExpr(xref)];
+  }
+  result = [e, e.clone()];
+  eliminateTemporaryAssignments(result[1], temporariesIn(result[0]), ctx);
```
At least a comment here about _why_ this happens

dylhunn commented:
caretaker: windows target is a flake. g3 is not affected.

alfaproject commented:
OOC are there plans to make these the same as JS semantics?

dylhunn commented:
@AndrewKushnir PR is rebased.
